### PR TITLE
1.3/1067 remove voucher from a bundle backend

### PR DESCRIPTION
**Pull request by charlesstrange2**
Should fulfil DELETE /registrations/{id}/vouchers/{id}
Some extra tests could be added.

### CODE REVIEW

File: app/Http/Controllers/Store/BundleController.php [hunk 1]
```diff
@@ -103,4 +83,62 @@ public function update(StoreUpdateBundleRequest $request, Registration $registra
...
+
+    public function removeVoucherFromCurrentBundle(Registration $registration, Voucher $voucher)
+    {
+        /** @var Bundle $bundle */
+        $bundle = $registration->currentBundle();
+
+        // It is attached to our bundle, right?
+        if ($voucher->bundle_id == $bundle->id) {
+            // Call alterVouchers with no codes, and no bundle to detransiton and remove it.
+            $errors = $bundle->alterVouchers(collect([$voucher]), [], null);
```
Comment (katjam) [inline app/Http/Controllers/Store/BundleController.php]: Fine to include codes and bundle params but these are already set as defaults.

File: app/Http/Controllers/Store/BundleController.php [hunk 2]
```diff
@@ -103,4 +83,62 @@ public function update(StoreUpdateBundleRequest $request, Registration $registra
...
+            // Error it out (how did you get here?
+            $errors["foreign"] = [$voucher->code];
+        }
+
+        if (!empty($errors)) {
+            $messages = $this->assembleMessages($errors);
+            // Spit the basic error messages back
+            return redirect()->route('store.registration.voucher-manager', ['registration' => $registration->id])
+                ->withInput()
+                ->with('error_message', join('. ', $messages) . '.');
```
Comment (katjam) [inline app/Http/Controllers/Store/BundleController.php]: Worth moving the join into assembleMessages?

File: routes/store.php [hunk 3]
```diff
@@ -67,21 +67,32 @@
...
+
     Route::get(
         '/registrations/{registration}/voucher-manager',
         'BundleController@create'
     )
     ->name('store.registration.voucher-manager')
     ->middleware('can:view,registration');
 
-    // puts (and replaces!) the currentBundle of vouchers!
+    // PUTS (and replaces!) the currentBundle of vouchers!
```
Comment (katjam) [inline routes/store.php]: Is the `POST registrations/{id}/vouchers [start, end]` coming in a later PR?

File: tests/Unit/Controllers/Store/BundleControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+use Tests\TestCase;
+
+class BundleControllerTest extends TestCase
+{
+    use DatabaseMigrations;
+
+    protected $centre;
+    protected $centreUser;
+
+    /** @var Registration $registration */
```
Comment (katjam) [inline tests/Unit/Controllers/Store/BundleControllerTest.php]: Are these intermittent doc comments being injected by your  IDE? there doesn't seem to be consistency of use.

File: tests/Unit/Controllers/Store/BundleControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        $this->centreUser =  factory(CentreUser::class)->create([
+            "name"  => "test user",
+            "email" => "testuser@example.com",
+            "password" => bcrypt('test_user_pass'),
+            "centre_id" => $this->centre->id,
+        ]);
+
+        //  A Registration on that centre
+        $this->registration = factory(Registration::class)->create([
+                'centre_id' => $this->centre->id
```
Comment (katjam) [inline tests/Unit/Controllers/Store/BundleControllerTest.php]: Indentation a bit off.

File: routes/store.php (hunk 3 above)
Comment (charlesstrange2) [reply routes/store.php]: Yes, this is just the delete endpoint

File: tests/Unit/Controllers/Store/BundleControllerTest.php (hunk 4 above)
Comment (charlesstrange2) [reply tests/Unit/Controllers/Store/BundleControllerTest.php]: Yes, it's phpdoc to tell the ide about objects it can't resolve.

**Pull request merged**